src/libmongoc/src/mongoc/mongoc-collection.c

```diff
@@ -750,27 +752,22 @@ mongoc_collection_count_with_opts (
    bool success;
    bson_t reply;
    bson_t cmd = BSON_INITIALIZER;
-   bson_t q;
 
    ENTRY;
 
    BSON_ASSERT_PARAM (collection);
 
-   bson_append_utf8 (
-      &cmd, "count", 5, collection->collection, collection->collectionlen);
-   if (query) {
-      bson_append_document (&cmd, "query", 5, query);
-   } else {
-      bson_init (&q);
-      bson_append_document (&cmd, "query", 5, &q);
-      bson_destroy (&q);
-   }
-   if (limit) {
-      bson_append_int64 (&cmd, "limit", 5, limit);
-   }
-   if (skip) {
-      bson_append_int64 (&cmd, "skip", 4, skip);
-   }
+   bsonBuildAppend (
+      cmd,
+      kv ("count",
+          utf8_w_len (collection->collection, collection->collectionlen)),
+      kv ("query",
+          if (query,                // If we have a query,
+              then (bson (*query)), // Copy it
+              else(doc ()))),       // Otherwise, add an empty doc
+      if (limit, then (kv ("limit", int64 (limit)))),
+      if (skip, then (kv ("skip", int64 (skip)))));
```

Comment:
```suggestion
   bsonBuildAppend (
      cmd,
      kvl ("count",
           5,
           utf8_w_len (collection->collection, collection->collectionlen)),
      kvl ("query",
           5,
           if (query,                // If we have a query,
               then (bson (*query)), // Copy it
               else(doc ()))),       // Otherwise, add an empty doc
      if (limit, then (kvl ("limit", 5, int64 (limit)))),
      if (skip, then (kvl ("skip", 4, int64 (skip)))));
```
Suggest preserving lengths to avoid calls to `strlen`.

Reply:
Resolving in response to https://github.com/mongodb/mongo-c-driver/pull/1094#discussion_r1004799519.